File: packages/sdk-py/src/leji/indexgen.py
"""Context index generation, currency checking, and serialization."""

from __future__ import annotations

import datetime as dt
import hashlib
import json
import posixpath
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Optional

from .findings import Finding
from .fsx import is_contained, resolved_within_root
from .gitutil import git_last_modified, git_toplevel
from .layer import duplicate_id_findings, read_json_artifact, scan_categories
from .manifest import Manifest, effective_index_path
from .schemas import SDK_VERSION, SUPPORTED_LINES, schema_errors

ID_PATTERN = re.compile(r"^[a-z0-9]+(-[a-z0-9]+)*$")

ENTRY_KEY_ORDER = [
    "id",
    "path",
    "title",
    "category",
    "kind",
    "date",
    "summary",
    "tags",
    "owners",
    "lastModified",
    "contentHash",
    "freshness",
    "links",
]

MOUNT_KEY_ORDER = ["path", "name", "owner", "role", "categories", "topics", "requiredWhen"]

_DATE = re.compile(r"^\d{4}-\d{2}-\d{2}$")


@dataclass
class IndexResult:
    index: Optional[dict]
    findings: list[Finding]
    stale: Optional[bool] = None


def _slugify(stem: str) -> str:
    return re.sub(r"^-+|-+$", "", re.sub(r"[^a-z0-9]+", "-", stem.lower()))


def _first_heading(body: str) -> Optional[str]:
    m = re.search(r"^#\s+(.+)$", body, re.MULTILINE)
    return m.group(1).strip() if m else None


def _content_hash(root: str, rel_path: str) -> str:
    digest = hashlib.sha256((Path(root) / rel_path).read_bytes()).hexdigest()
    return "sha256:" + digest[:16]


def _str(v: Any) -> Optional[str]:
    return v if isinstance(v, str) and v != "" else None


def _str_array(v: Any) -> Optional[list[str]]:
    if not isinstance(v, list):
        return None
    out = [x for x in v if isinstance(x, str)]
    return out or None


def load_stored_index(root: str, manifest: Manifest) -> Optional[dict]:
    rel = effective_index_path(manifest)
    abs_path = Path(root) / rel
    # Refuse an artifact whose real path escapes the repo root (e.g. a symlinked
    # index pointing outside the tree); a layer's index lives inside the layer.
    if not abs_path.is_file() or not is_contained(root, abs_path):
        return None
    data, _ = read_json_artifact(root, rel)
    return data if isinstance(data, dict) else None


def generate_index(root: str, manifest: Manifest) -> IndexResult:
    """Generate the context index from the tree.

    Id stability, in priority order: document frontmatter ``id``, the stored
    index's id for the same path, the stored index's id for the same
    contentHash (a pure move), then a filename slug (de-collided with the
    parent directory).
    """
    findings: list[Finding] = []
    scan = scan_categories(root, manifest)
    findings.extend(scan.findings)
    docs = scan.docs
    stored = load_stored_index(root, manifest)
    stored_by_path: dict[str, dict] = {}
    # Carry an id by content-hash only when that hash maps to exactly one stored
    # entry: two byte-identical documents share a hash, so a hash-carry there would
    # misattribute one document's id to the other on a move.
    hash_entries: dict[str, list[dict]] = {}
    for stored_entry in (stored or {}).get("entries", []):
        stored_by_path[stored_entry.get("path", "")] = stored_entry
        if stored_entry.get("contentHash"):
            hash_entries.setdefault(stored_entry["contentHash"], []).append(stored_entry)
    stored_by_hash: dict[str, dict] = {
        h: arr[0] for h, arr in hash_entries.items() if len(arr) == 1
    }

    in_git = git_toplevel(root) is not None
    today = dt.datetime.now(dt.timezone.utc).date().isoformat()
    used: dict[str, str] = {}
    entries: list[dict] = []

    for doc in docs:
        fm = doc.frontmatter or {}
        content_hash = _content_hash(root, doc.rel_path)
        carried = stored_by_path.get(doc.rel_path) or stored_by_hash.get(content_hash)

        entry_id = _str(fm.get("id")) or (carried or {}).get("id")
        if not entry_id:
            stem = posixpath.basename(doc.rel_path).removesuffix(".md")
            entry_id = _slugify(stem)
            if entry_id in used:
                parent = _slugify(posixpath.basename(posixpath.dirname(doc.rel_path)))
                entry_id = f"{parent}-{entry_id}" if parent else entry_id
            candidate, n = entry_id, 2
            while candidate in used:
                candidate = f"{entry_id}-{n}"
                n += 1
            entry_id = candidate
        if not ID_PATTERN.fullmatch(entry_id):
            findings.append(
                Finding(
                    "id-pattern",
                    "error",
                    f'derived id "{entry_id}" is not lowercase-hyphen',
                    doc.rel_path,
                )
            )
        if entry_id in used:
            findings.append(
                Finding(
                    "id-duplicate",
                    "error",
                    f'index id "{entry_id}" already used by {used[entry_id]}',
                    doc.rel_path,
                )
            )
        used[entry_id] = doc.rel_path

        entry: dict = {
            "id": entry_id,
            "path": doc.rel_path,
            "title": _str(fm.get("title"))
            or _first_heading(doc.body)
            or posixpath.basename(doc.rel_path).removesuffix(".md"),
            "category": doc.category,
            "kind": doc.kind,
        }
        if doc.kind == "record":
            # A record's date comes only from explicit, valid frontmatter; nothing
            # is scraped from prose or filename conventions.
            date = _str(fm.get("date"))
            if date and _DATE.fullmatch(date):
                entry["date"] = date
        summary = _str(fm.get("summary")) or (carried or {}).get("summary")
        if summary:
            entry["summary"] = summary
        tags = _str_array(fm.get("tags"))
        if tags:
            entry["tags"] = tags
        owners = _str_array(fm.get("owners"))
        if owners:
            entry["owners"] = owners
        entry["lastModified"] = (git_last_modified(root, doc.rel_path) if in_git else None) or today
        entry["contentHash"] = content_hash
        freshness = fm.get("freshness") or {}
        review_after = _str(freshness.get("reviewAfter")) if isinstance(freshness, dict) else None
        if review_after:
            entry["freshness"] = {"reviewAfter": review_after}
        links = _str_array(fm.get("links"))
        if links:
            entry["links"] = links
        entries.append(entry)

    # Id churn: a stored id whose path is gone and that did not reappear at a new path
    # vanished (a document moved AND edited with no frontmatter id mints a fresh slug).
    # Inbound references to the old id now dangle; warn so it is caught, not silent.
    new_ids = {e["id"] for e in entries}
    current_paths = {d.rel_path for d in docs}
    for stored_entry in (stored or {}).get("entries", []):
        p = stored_entry.get("path", "")
        sid = stored_entry.get("id", "")
        if p not in current_paths and sid not in new_ids:
            findings.append(
                Finding(
                    "id-vanished",
                    "warning",
                    f'stored id "{sid}" (was {p}) did not reappear; references to it now dangle. '
                    "Declare a frontmatter id to keep ids stable across moves.",
                    p,
                )
            )

    mounts: list[dict] = []
    federation = manifest.get("federation") or {}
    for mt in federation.get("mounts") or []:
        rec: dict = {"name": mt["name"], "source": mt["source"], "pin": mt["pin"]}
        if mt.get("trackingRef"):
            rec["trackingRef"] = mt["trackingRef"]
        rec["owner"] = mt["owner"]
        if mt.get("role"):
            rec["role"] = mt["role"]
        if mt.get("categories"):
            rec["categories"] = mt["categories"]
        if mt.get("topics"):
            rec["topics"] = mt["topics"]
        if mt.get("requiredWhen"):
            rec["requiredWhen"] = mt["requiredWhen"]
        mounts.append(rec)

    index = {
        "$schema": "https://leji.org/schemas/v1.0/context-index.schema.json",
        "schemaVersion": "1.0",
        "generatedAt": dt.datetime.now(dt.timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z",
        "generator": {"name": "leji", "version": SDK_VERSION},
        "rootPath": manifest["rootPath"],
        "entries": entries,
    }
    if mounts:
        index["mounts"] = mounts
    return IndexResult(index=index, findings=findings)


def _comparable(entry: dict) -> dict:
    """Currency-comparison view of an entry; volatile fields excluded."""
    return {k: v for k, v in entry.items() if k != "lastModified"}


def _stable_stringify(value: object) -> str:
    """Key-order-insensitive serialization, mirrored by the Node SDK."""
    return json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def check_index(root: str, manifest: Manifest) -> IndexResult:
    """Check the stored index against a regeneration.

    ``generatedAt``, ``generator``, and ``lastModified`` are excluded from the
    comparison: content drift is what ``contentHash`` catches deterministically.
    """
    rel = effective_index_path(manifest)
    findings: list[Finding] = []
    if not (Path(root) / rel).is_file():
        findings.append(
            Finding(
                "index-required",
                "error",
                f"index {rel} does not exist; run `leji index`",
                rel,
            )
        )
        return IndexResult(index=None, findings=findings, stale=True)
    if not is_contained(root, Path(root) / rel):
        findings.append(
            Finding(
                "artifact-parse", "error", f"artifact {rel} resolves outside the layer root", rel
            )
        )
        return IndexResult(index=None, findings=findings, stale=True)

    stored = load_stored_index(root, manifest)
    if stored is None:
        findings.append(Finding("artifact-parse", "error", "stored index is not valid JSON", rel))
        return IndexResult(index=None, findings=findings, stale=True)
    for err in schema_errors("context-index", stored):
        findings.append(Finding("artifact-schema", "error", err, rel))
    stored_version = stored.get("schemaVersion")
    if isinstance(stored_version, str) and stored_version not in SUPPORTED_LINES:
        findings.append(
            Finding(
                "schema-version",
                "error",
                f'schemaVersion "{stored_version}" is not supported by this SDK',
                rel,
            )
        )
    if findings:
        return IndexResult(index=stored, findings=findings, stale=True)

    regen = generate_index(root, manifest)
    assert regen.index is not None  # generate_index always produces an index
    # A regeneration that itself errors (missing/malformed index file, a
    # category-conflict, a dangling entry) means the tree cannot be indexed
    # cleanly, so the stored index cannot be current: fail rather than compare a
    # partial regen against it and falsely pass.
    regen_errors = [f for f in regen.findings if f.severity == "error"]
    if regen_errors:
        findings.extend(regen_errors)
        return IndexResult(index=stored, findings=findings, stale=True)
    want = _stable_stringify(
        {
            "rootPath": regen.index["rootPath"],
            "entries": [_comparable(e) for e in regen.index["entries"]],
            "mounts": regen.index.get("mounts", []),
        }
    )
    got = _stable_stringify(
        {
            "rootPath": stored.get("rootPath"),
            "entries": [
                _comparable(e)
                for e in sorted(stored.get("entries", []), key=lambda e: e.get("path", ""))
            ],
            "mounts": stored.get("mounts", []),
        }
    )
    if want != got:
        want_paths = {e["path"] for e in regen.index["entries"]}
        got_paths = {e.get("path") for e in stored.get("entries", [])}
        missing = len(want_paths - got_paths)
        extra = len(got_paths - want_paths)
        detail = (
            f" (missing: {missing}, removed: {extra})"
            if (missing or extra)
            else " (entry content drifted)"
        )
        findings.append(
            Finding(
                "index-stale",
                "error",
                f"index no longer matches the tree{detail}; run `leji index`",
                rel,
            )
        )
        return IndexResult(index=stored, findings=findings, stale=True)
    dup = duplicate_id_findings(
        [(e.get("id"), e.get("path", "")) for e in stored.get("entries", [])], "index"
    )
    return IndexResult(index=stored, findings=[*regen.findings, *dup], stale=False)


def _ordered_mount(m: dict) -> dict:
    out: dict = {}
    for key in MOUNT_KEY_ORDER:
        if key not in m:
            continue
        if key == "owner":
            owner = m["owner"]
            owner_out = {"name": owner["name"]}
            if owner.get("contact") is not None:
                owner_out["contact"] = owner["contact"]
            out["owner"] = owner_out
        else:
            out[key] = m[key]
    return out


def serialize_index(index: dict) -> str:
    """Stable key order, 2-space indent, trailing newline."""
    out = {
        "$schema": index.get("$schema"),
        "schemaVersion": index["schemaVersion"],
        "generatedAt": index["generatedAt"],
        "generator": index.get("generator"),
        "rootPath": index["rootPath"],
        "entries": [{key: e[key] for key in ENTRY_KEY_ORDER if key in e} for e in index["entries"]],
    }
    if index.get("mounts"):
        out["mounts"] = [_ordered_mount(m) for m in index["mounts"]]
    return json.dumps(out, indent=2, ensure_ascii=False) + "\n"


def write_index(root: str, manifest: Manifest) -> IndexResult:
    rel = effective_index_path(manifest)
    result = generate_index(root, manifest)
    # Refuse to write a partial or incorrect index when generation hit a hard
    # error (e.g. category-conflict, index-file-parse, a dangling entry): writing
    # would persist a half-correct artifact that later reads trust.
    if any(f.severity == "error" for f in result.findings):
        return result
    if result.index is not None:
        abs_path = Path(root) / rel
        # Contain before creating any directory: resolved_within_root resolves the
        # nearest existing ancestor, so a symlinked ancestor of this not-yet-existing
        # target is caught before mkdir/write can escape the layer root.
        if not resolved_within_root(root, abs_path):
            return IndexResult(
                index=result.index,
                findings=[
                    *result.findings,
                    Finding(
                        "artifact-parse",
                        "error",
                        f"index path {rel} resolves outside the layer root",
                        rel,
                    ),
                ],
            )
        abs_path.parent.mkdir(parents=True, exist_ok=True)
        abs_path.write_text(serialize_index(result.index), encoding="utf-8")
    return result

File: packages/sdk/src/commands/indexgen.ts
import * as crypto from 'node:crypto';
import * as fs from 'node:fs';
import * as path from 'node:path';
import { type Finding, finding } from '../lib/findings.js';
import { guardRoot, isFile, resolvedWithinRoot, verifiedTargetRead, writeFileGuarded } from '../lib/fsx.js';
import { gitLastModified, gitToplevel } from '../lib/git.js';
import { duplicateIdFindings, scanCategories } from '../lib/layer.js';
import { type Manifest, effectiveIndexPath } from '../lib/manifest.js';
import { SDK_VERSION, SUPPORTED_LINES, schemaErrors } from '../lib/schemas.js';

/** One artifact's entry in the generated context index. */
export interface IndexEntry {
   id: string;
   path: string;
   title: string;
   category: string;
   /** The document's kind. Always emitted; consumers treat an absent
    * value (older indexes) as `intent`. */
   kind: 'intent' | 'record';
   /** A record's date, sourced only from valid frontmatter `date`. */
   date?: string;
   summary?: string;
   tags?: string[];
   owners?: string[];
   lastModified?: string;
   contentHash?: string;
   freshness?: { reviewAfter: string };
   links?: string[];
}

/** One federated sibling's routing record in the generated index, derived from
 * the manifest's federation.mounts. Routing metadata only, never sibling content. */
export interface IndexMount {
   name: string;
   source: string;
   pin: string;
   trackingRef?: string;
   owner: { name: string; contact?: string };
   role?: string;
   categories?: string[];
   topics?: string[];
   requiredWhen?: string[];
}

/** The generated machine index: every indexed artifact in the layer. */
export interface ContextIndex {
   $schema?: string;
   schemaVersion: string;
   generatedAt: string;
   generator?: { name?: string; version?: string };
   rootPath: string;
   entries: IndexEntry[];
   mounts?: IndexMount[];
}

export interface IndexResult {
   index: ContextIndex | null;
   findings: Finding[];
   /** Set by check(): true when the stored index no longer matches the tree. */
   stale?: boolean;
}

const ID_PATTERN = /^[a-z0-9]+(-[a-z0-9]+)*$/;

function slugify(stem: string): string {
   return stem
      .toLowerCase()
      .replace(/[^a-z0-9]+/g, '-')
      .replace(/^-+|-+$/g, '');
}

function firstHeading(body: string): string | null {
   const m = /^#\s+(.+)$/m.exec(body);
   return m ? m[1].trim() : null;
}

function contentHash(root: string, relPath: string): string {
   const buf = fs.readFileSync(path.join(root, relPath));
   return 'sha256:' + crypto.createHash('sha256').update(buf).digest('hex').slice(0, 16);
}

function str(v: unknown): string | undefined {
   return typeof v === 'string' && v !== '' ? v : undefined;
}

function strArray(v: unknown): string[] | undefined {
   if (!Array.isArray(v)) return undefined;
   const out = v.filter((x): x is string => typeof x === 'string');
   return out.length > 0 ? out : undefined;
}

/** The stored index, or null when there is none this run can act on. Read through
 * the verified read, not by pathname: generation carries ids out of these bytes into
 * the index it writes back to this same path, so the file that was judged must be the
 * file that is read. Absent, unparsable, or a standing entry that cannot be verified
 * all mean "no stored index" — nothing is carried, and the write chokepoint judges the
 * destination again on its own. */
export function loadStoredIndex(root: string, manifest: Manifest): ContextIndex | null {
   const rel = effectiveIndexPath(manifest);
   const read = verifiedTargetRead(guardRoot(root), path.join(root, rel), null);
   if (read.status !== 'regular') return null;
   let data: unknown;
   try {
      data = JSON.parse(read.bytes.toString('utf8'));
   } catch {
      return null;
   }
   if (!data || typeof data !== 'object') return null;
   return data as ContextIndex;
}

/**
 * Generate the context index from the tree. Id stability priority: frontmatter
 * `id`, stored id for the same path, stored id for the same contentHash (a pure
 * move), then a filename slug (de-collided with the parent directory).
 */
export function generateIndex(root: string, manifest: Manifest): IndexResult {
   const findings: Finding[] = [];
   const scan = scanCategories(root, manifest);
   findings.push(...scan.findings);
   const docs = scan.docs;
   const stored = loadStoredIndex(root, manifest);
   const storedByPath = new Map<string, IndexEntry>();
   // Carry an id by content-hash only when the hash maps to exactly one stored entry:
   // byte-identical documents share a hash, so a carry would misattribute an id.
   const hashEntries = new Map<string, IndexEntry[]>();
   for (const entry of stored?.entries ?? []) {
      storedByPath.set(entry.path, entry);
      if (entry.contentHash) {
         const arr = hashEntries.get(entry.contentHash) ?? [];
         arr.push(entry);
         hashEntries.set(entry.contentHash, arr);
      }
   }
   const storedByHash = new Map<string, IndexEntry>();
   for (const [h, arr] of hashEntries) if (arr.length === 1) storedByHash.set(h, arr[0]);

   const inGit = gitToplevel(root) !== null;
   const today = new Date().toISOString().slice(0, 10);
   const used = new Map<string, string>();
   const entries: IndexEntry[] = [];

   for (const doc of docs) {
      const fm = doc.frontmatter ?? {};
      const hash = contentHash(root, doc.relPath);
      const carried = storedByPath.get(doc.relPath) ?? storedByHash.get(hash);

      let id = str(fm.id) ?? carried?.id;
      if (!id) {
         const stem = path.posix.basename(doc.relPath).replace(/\.md$/, '');
         id = slugify(stem);
         if (used.has(id)) {
            const parent = slugify(path.posix.basename(path.posix.dirname(doc.relPath)));
            id = parent ? `${parent}-${id}` : id;
         }
         let candidate = id;
         let n = 2;
         while (used.has(candidate)) candidate = `${id}-${n++}`;
         id = candidate;
      }
      if (!ID_PATTERN.test(id)) {
         findings.push(finding('id-pattern', 'error', `derived id "${id}" is not lowercase-hyphen`, doc.relPath));
      }
      if (used.has(id)) {
         findings.push(
            finding('id-duplicate', 'error', `index id "${id}" already used by ${used.get(id)}`, doc.relPath),
         );
      }
      used.set(id, doc.relPath);

      const entry: IndexEntry = {
         id,
         path: doc.relPath,
         title: str(fm.title) ?? firstHeading(doc.body) ?? path.posix.basename(doc.relPath).replace(/\.md$/, ''),
         category: doc.category,
         kind: doc.kind,
      };
      if (doc.kind === 'record') {
         // A record's date comes only from explicit, valid frontmatter; nothing
         // is scraped from prose or filename conventions.
         const d = str(fm.date);
         if (d && /^\d{4}-\d{2}-\d{2}$/.test(d)) entry.date = d;
      }
      const summary = str(fm.summary) ?? carried?.summary;
      if (summary) entry.summary = summary;
      const tags = strArray(fm.tags);
      if (tags) entry.tags = tags;
      const owners = strArray(fm.owners);
      if (owners) entry.owners = owners;
      entry.lastModified = (inGit ? gitLastModified(root, doc.relPath) : null) ?? today;
      entry.contentHash = hash;
      const freshness = fm.freshness as { reviewAfter?: unknown } | undefined;
      const reviewAfter = str(freshness?.reviewAfter);
      if (reviewAfter) entry.freshness = { reviewAfter };
      const links = strArray(fm.links);
      if (links) entry.links = links;
      entries.push(entry);
   }

   // Id churn: a stored id whose path is gone and that didn't reappear has vanished
   // (move + edit in one change set with no frontmatter `id` mints a fresh slug).
   // Inbound references now dangle; warn. A frontmatter `id` makes ids move-proof.
   const newIds = new Set(entries.map((e) => e.id));
   const currentPaths = new Set(docs.map((d) => d.relPath));
   for (const entry of stored?.entries ?? []) {
      if (!currentPaths.has(entry.path) && !newIds.has(entry.id)) {
         findings.push(
            finding(
               'id-vanished',
               'warning',
               `stored id "${entry.id}" (was ${entry.path}) did not reappear; references to it now dangle. Declare a frontmatter id to keep ids stable across moves.`,
               entry.path,
            ),
         );
      }
   }

   const mounts: IndexMount[] = (manifest.federation?.mounts ?? []).map((m) => {
      const rec: IndexMount = { name: m.name, source: m.source, pin: m.pin, owner: m.owner };
      if (m.trackingRef) rec.trackingRef = m.trackingRef;
      if (m.role) rec.role = m.role;
      if (m.categories && m.categories.length > 0) rec.categories = m.categories;
      if (m.topics && m.topics.length > 0) rec.topics = m.topics;
      if (m.requiredWhen && m.requiredWhen.length > 0) rec.requiredWhen = m.requiredWhen;
      return rec;
   });

   const index: ContextIndex = {
      $schema: 'https://leji.org/schemas/v1.0/context-index.schema.json',
      schemaVersion: '1.0',
      generatedAt: new Date().toISOString(),
      generator: { name: 'leji', version: SDK_VERSION },
      rootPath: manifest.rootPath,
      entries,
   };
   if (mounts.length > 0) index.mounts = mounts;
   return { index, findings };
}

/** Fields compared for currency; volatile fields are deliberately excluded. */
function comparable(entry: IndexEntry): Omit<IndexEntry, 'lastModified'> {
   const { lastModified: _lastModified, ...rest } = entry;
   return rest;
}

/** Key-order-insensitive serialization, mirrored by the Python SDK. */
export function stableStringify(value: unknown): string {
   if (Array.isArray(value)) return `[${value.map(stableStringify).join(',')}]`;
   if (value !== null && typeof value === 'object') {
      const keys = Object.keys(value as Record<string, unknown>).sort();
      return `{${keys
         .map((k) => `${JSON.stringify(k)}:${stableStringify((value as Record<string, unknown>)[k])}`)
         .join(',')}}`;
   }
   return JSON.stringify(value);
}

/**
 * Check the stored index against a regeneration. `generatedAt`, `generator`, and
 * `lastModified` are excluded; `contentHash` catches drift deterministically.
 */
export function checkIndex(root: string, manifest: Manifest): IndexResult {
   const rel = effectiveIndexPath(manifest);
   const findings: Finding[] = [];
   if (!isFile(path.join(root, rel))) {
      findings.push(finding('index-required', 'error', `index ${rel} does not exist; run \`leji index\``, rel));
      return { index: null, findings, stale: true };
   }
   if (!resolvedWithinRoot(path.resolve(root), path.join(root, rel))) {
      findings.push(finding('artifact-parse', 'error', `artifact ${rel} resolves outside the layer root`, rel));
      return { index: null, findings, stale: true };
   }

   const stored = loadStoredIndex(root, manifest);
   if (!stored) {
      findings.push(finding('artifact-parse', 'error', 'stored index is not valid JSON', rel));
      return { index: null, findings, stale: true };
   }
   for (const err of schemaErrors('context-index', stored)) {
      findings.push(finding('artifact-schema', 'error', err, rel));
   }
   if (typeof stored.schemaVersion === 'string' && !SUPPORTED_LINES.includes(stored.schemaVersion)) {
      findings.push(
         finding(
            'schema-version',
            'error',
            `schemaVersion "${stored.schemaVersion}" is not supported by this SDK`,
            rel,
         ),
      );
   }
   if (findings.length > 0) return { index: stored, findings, stale: true };

   const regen = generateIndex(root, manifest);
   // If regeneration itself errors, the tree can't be indexed cleanly, so the stored
   // index can't be current: fail rather than compare a partial regen and falsely pass.
   const regenErrors = regen.findings.filter((f) => f.severity === 'error');
   if (regenErrors.length > 0) {
      findings.push(...regenErrors);
      return { index: stored, findings, stale: true };
   }
   const want = stableStringify({
      rootPath: regen.index!.rootPath,
      entries: regen.index!.entries.map(comparable),
      mounts: regen.index!.mounts ?? [],
   });
   const got = stableStringify({
      rootPath: stored.rootPath,
      entries: [...stored.entries].sort((a, b) => (a.path < b.path ? -1 : 1)).map(comparable),
      mounts: stored.mounts ?? [],
   });
   if (want !== got) {
      const wantPaths = new Set(regen.index!.entries.map((e) => e.path));
      const gotPaths = new Set(stored.entries.map((e) => e.path));
      const missing = [...wantPaths].filter((p) => !gotPaths.has(p));
      const extra = [...gotPaths].filter((p) => !wantPaths.has(p));
      const detail =
         missing.length > 0 || extra.length > 0
            ? ` (missing: ${missing.length}, removed: ${extra.length})`
            : ' (entry content drifted)';
      findings.push(
         finding('index-stale', 'error', `index no longer matches the tree${detail}; run \`leji index\``, rel),
      );
      return { index: stored, findings, stale: true };
   }
   return {
      index: stored,
      findings: [
         ...regen.findings,
         ...duplicateIdFindings(
            stored.entries.map((e) => ({ id: e.id, relPath: e.path })),
            'index',
         ),
      ],
      stale: false,
   };
}

const ENTRY_KEY_ORDER: (keyof IndexEntry)[] = [
   'id',
   'path',
   'title',
   'category',
   'kind',
   'date',
   'summary',
   'tags',
   'owners',
   'lastModified',
   'contentHash',
   'freshness',
   'links',
];

function orderedEntry(entry: IndexEntry): Record<string, unknown> {
   const out: Record<string, unknown> = {};
   for (const key of ENTRY_KEY_ORDER) {
      if (entry[key] !== undefined) out[key] = entry[key];
   }
   return out;
}

const MOUNT_KEY_ORDER: (keyof IndexMount)[] = [
   'name',
   'source',
   'pin',
   'trackingRef',
   'owner',
   'role',
   'categories',
   'topics',
   'requiredWhen',
];

function orderedMount(m: IndexMount): Record<string, unknown> {
   const out: Record<string, unknown> = {};
   for (const key of MOUNT_KEY_ORDER) {
      if (m[key] === undefined) continue;
      if (key === 'owner') {
         const o: Record<string, unknown> = { name: m.owner.name };
         if (m.owner.contact !== undefined) o.contact = m.owner.contact;
         out.owner = o;
      } else {
         out[key] = m[key];
      }
   }
   return out;
}

/** Serialize an index with stable key order, 2-space indent, trailing newline. */
export function serializeIndex(index: ContextIndex): string {
   const out: Record<string, unknown> = {
      $schema: index.$schema,
      schemaVersion: index.schemaVersion,
      generatedAt: index.generatedAt,
      generator: index.generator,
      rootPath: index.rootPath,
      entries: index.entries.map(orderedEntry),
   };
   if (index.mounts && index.mounts.length > 0) out.mounts = index.mounts.map(orderedMount);
   return JSON.stringify(out, null, 2) + '\n';
}

/** Generate and write the index to the effective path. */
export function writeIndex(root: string, manifest: Manifest): IndexResult {
   const rel = effectiveIndexPath(manifest);
   const result = generateIndex(root, manifest);
   // Refuse to write when generation hit a hard error: persisting a half-correct
   // index would be trusted by later reads.
   if (result.findings.some((f) => f.severity === 'error')) {
      return result;
   }
   if (result.index) {
      const abs = path.join(root, rel);
      // The write chokepoint judges the RESOLVED destination immediately before the
      // write, catching a symlinked ancestor before anything is created under it.
      if (!writeFileGuarded(guardRoot(root), abs, null, serializeIndex(result.index)).ok) {
         return {
            index: result.index,
            findings: [
               ...result.findings,
               finding('artifact-parse', 'error', `index path ${rel} resolves outside the layer root`, rel),
            ],
         };
      }
   }
   return result;
}
